courseworks/week4/xuanzhaogao/hw4.ipynb

```diff
@@ -0,0 +1,269 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Question 1\n",
+    "   1. four householder transformations are required.\n",
+    "   2. the first column of $A$ after transformation will be $(2, 0, 0, 0)$.\n",
+    "   3. the second transformation will not affect the first column of $A$.\n",
+    "   4. $6$ times of Givens rotations would be required."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "using LinearAlgebra"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 119,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "Givens_rotation (generic function with 1 method)"
+      ]
+     },
+     "execution_count": 119,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "function Givens_rotation(A)\n",
+    "    n = size(A, 1)\n",
+    "    Q = Matrix{Float64}(I, n, n)\n",
+    "    R = copy(A)\n",
+    "    for i = 1:n - 1\n",
+    "        t_ii = R[i, i]\n",
+    "        t_ji = R[i + 1, i]\n",
+    "\n",
+    "        r = (t_ii^2 + t_ji^2)^.5\n",
+    "        c = t_ii / r\n",
+    "        s = t_ji / r\n",
+    "\n",
+    "\n",
+    "        for j in i:n\n",
+    "            A_1 = R[i, j]\n",
+    "            A_2 = R[i + 1, j]\n",
+    "            R[i, j] = c * A_1 + s * A_2\n",
+    "            R[i + 1, j] = - s * A_1 + c * A_2\n",
+    "        end\n",
+    "        \n",
+    "        for j in 1:n\n",
+    "            Q_1 = Q[j, i]\n",
+    "            Q_2 = Q[j, i + 1]\n",
+    "            Q[j, i] = c * Q_1 + s * Q_2\n",
+    "            Q[j, i + 1] = - s * Q_1 + c * Q_2\n",
+    "        end\n",
+    "    end\n",
+    "\n",
+    "    return Q, R\n",
+    "end"
```

Comment:
Two extra comments:
1. Defining a function for applying the givens rotation matrix will improve the code readability.
2. Adding `@inbounds` before the for-loop will improve the performance.